### ISSUE TITLE
Fix deform produce incorrect restarts

### ISSUE THREAD

**Issue by antonpeshkov** [lammps/lammps#2944]
This is a continuation of the bug report described in the following MATSCI thread https://matsci.org/t/fix-deform-screw-restarts-and-what-to-do-with-it/38239 . 

They are two bugs related to fix deform and restarts. While preparing the report for this bug, I discovered two more problems, one related to the Windows build of lammps and another to the virial computation for spherical particles. As these two bugs are not related to fix deform, I will open separate issues for them. 

First, because box size/shape change happens after force computation, then if a run is restarted, the newly computed forces will be different from those computed by lammps before the restart.

Here is the archive, with files needed to reproduce this error:

[Lammps bug report.zip](https://github.com/lammps/lammps/files/7157474/Lammps.bug.report.zip)

The “*_random_positons” files contain the initial positions of particles. There are two lammps scripts that are identical, except that one use the lj/cut pair style, as asked by Alex, and the other the granular pair style. The script performs an isotropic compression with a true strain rate for N steps. It then makes a “run 0” to recompute the pressure as if we had restarted the simulation. At the end it writes to the screen the pressure after the N steps and the pressure after the “0” step restart. You can see than in both of the cases of lj and granular potentials these pressures are close but not equal (N.B. do not use Windows builds to explore these bugs, as mentioned previously, they are broken). However, in some cases, they can be wildly different.

To illustrate what is happening I include a Julia script that computes the pressure from the dump file output by the granular script. Initially the granular lammps script use no velocity-dependent forces (gamman=0). If you run with Julia the “collisional_damping_pressure_test.jl” file, it will compute you the pressure from the dump file that is equal to the one computed by lammps after the restart. Now, if you uncomment lines 45-47 in the Julia script, it will reincrease the box size one step back before computing the pressure. In that case, it will find a pressure equivalent to the one computed by lammps at the end of N steps. Note, that in order to normalize the pressure by the volume, I still use the volume after the final box compaction. This makes the behavior of pressure computation and fix deform even more nonsensical, as the forces are computed at one volume, but then normalized at another volume to compute the pressure.

I think the best solution to all of the previously mentioned problems would be to move the box shape change in fix deform from fix->end_of_step() to fix->pre_exchange(). It will automatically resolve this bug. I actually do not understand why it has not been implemented that way since the beginning as it makes much more sense. In the current configuration, the box shape will be changed at step 1, the force corresponding to these shape changes will be computed at step 2 and the atom position motion in response to the shape change at step 1 will happen only at step 3… But I understand that making such a change may imply changing a lot of other lammps code, and may be impractical. 

The second bug concerns only the granular pair styles, in which the force can depend on the relative velocity of the particles. When lammps restart a simulation, fix deform does not remap the velocity of ghost particles for the initial force computation. This leads to greatly overestimated forces on the border particles.

To illustrate the second bug, go to line 20 in the granular lammps script and set gn=1. We now have forces that depend on the relative velocities of the particles. Run lammps again on this changed script and you will see that now the two pressures are completely different. In the Julia script, comment back lines 45-47 and in “contacts_stresses_collisional_damping_v_substract_hooke.jl” set gn=1 on line 46. Compute the pressure with Julia and you will see that once again it is equal to the one computed by lammps after the restart. Now uncomment back lines 45-47 in the main Julia file and remove the comments on lines 135-136 of the secondary Julia file. This will remove the deformation velocity from the ghost particles (i.e. will remap the velocity of ghost particles). Now you will see that the pressure is equal to the one before restart (there is some small difference due to the half-step difference in velocity between what lammps used to compute the pressure and the one saved to the files).

I assume that fixing this second bug should be straightforward.

**Comment by antonpeshkov** [comment]
A correction, what I thought was a Windows build bug, is actually none. I totally forgot that I modified my linux version of lammps to output the dump files with full double precision. By default the write_dump command use single precision output. Therefore, if you want to compare lammps output and the values computed by my Julia script you need to modify write_dump to output with double precision. Else, they will be a non-negligible difference in values between the ones computed by lammps and the ones computed by the Julia script.

**Comment by jtclemm** [comment]
> First, because box size/shape change happens after force computation, then if a run is restarted, the newly computed forces will be different from those computed by lammps before the restart.

Slight inconsistencies in forces across restarts are produced by many other features in LAMMPS. E.g.: pairstyles that depend on velocity (granular, SPH, DPD, etc.), some fixes that apply forces (fix viscous), and other fixes that adjust box dimensions (fix nh, berendsen, etc.). If it's important to ensure forces are consistent across restart files, I would argue it would be better to (optionally) save forces to restart files rather than playing whack-a-mole anytime an inconsistency is found.

> I think the best solution to all of the previously mentioned problems would be to move the box shape change in fix deform from fix->end_of_step() to fix->pre_exchange(). 

While this may fix the reported issue, it would break the `remap x` option.

> The second bug concerns only the granular pair styles, in which the force can depend on the relative velocity of the particles. When lammps restart a simulation, fix deform does not remap the velocity of ghost particles for the initial force computation. This leads to greatly overestimated forces on the border particles.

This appears to be due to the fact that `FixDeform->init()` zeros the `domain->h_rate` array. I don't see why this is necessary and the bug could be resolved by alternatively zeroing `domain->h_rate` in the constructor (and saving `domain->h_rate` to restart files). That way, fix deform would maintain memory of the most recent strain rate until it is deleted. @sjplimp Does this seem correct?

**Comment by sjplimp** [comment]
>> This appears to be due to the fact that FixDeform->init() zeros the domain->h_rate array. I don't see why this is necessary and the bug could be resolved by alternatively zeroing domain->h_rate in the constructor (and saving domain->h_rate to restart files). That way, fix deform would maintain memory of the most recent strain rate until it is deleted. @sjplimp Does this seem correct?

This seems like a good idea.  Are there any of the deformation modes where this would not be a good idea?
Also, there is no guarantee that when a user specifies a new fix deform command in a restart script, that
that they will specify it "correctly" to "continue" a previous deformation cycle, or even that that behavior is their intent.
In that scenario would it be "incorrect" to preserve the prior h_rate matrix?  If so, maybe there should an
added fix deform flag for the user to request that h_rate be preserved, otherwise don't do it.
OTOH, I think the criteria to re-use restart info in a new fix deform is that the fix-ID also match.  So maybe
emphasizing this issue in the doc page is a sufficient warning.

**Comment by jtclemm** [comment]
> In that scenario would it be "incorrect" to preserve the prior h_rate matrix?

Yes, in that instance ghost atoms would initially assuming an incorrect old h_rate matrix instead of assuming an incorrect zeroed matrix. Since the deformation probably produced some inertia in the system, having h_rate persist for that initial step might not be terrible. 

Fix deform is also only restarted if the style is consistent: https://github.com/lammps/lammps/blob/1144a01658fe1088179f1bf70a285e61165efe01/src/fix_deform.cpp#L961

So this should minimize the number of times there's a mismatch on restarting. 

Possibly a more problematic issue is what happens when one uses a deformation style that relies on the initial box dimensions, e.g. xy erate 1e-3. Here the rate that the angle of the box changes is calculated based on the initial perpendicular box height. So if the perpendicular height changes (via another deformation command or a change_box command), then the h_rate will assume the old box height. Although, again, the question is whether this is a worse assumption than h_rate[*] = 0.0.


> added fix deform flag for the user to request that h_rate be preserved, otherwise don't do it.

That might not be a bad option, although the implications of such a flag might be lost since it probably is just a tiny numerical effect in 99% of cases.

Alternatively, what if the code that calculates h_rate in the end_of_step() method was duplicated into init()? This would make sure ghosts have the value of h_rate that is about to be calculated when a second run command is invoked or a system is restarted. However, it would change the default behavior of LAMMPS since fix deform would then behave differently when it is first used (h_rate != 0.0 immediately).
